Update empty_loop documentation/message.

josephlr opened the pull request:
Originally part of #6161, but now this PR only deals with `std` crates

This change:
  - Updates the `std` message .
  - Updates the docs to mention how the busy loops should be fixed
    - Gives examples of how to do this for `no_std` targets
  - Updates the tests/stderr files to test this change.

changelog: Update `empty_loop` lint documentation

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @phansch (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust-clippy/blob/master/CONTRIBUTING.md) for more information.

phansch commented:
Thanks for the PR! The implementation looks good to me but I'm not using no_std myself, so I've asked on Zulip for a second pair of eyes.

Lokathor commented:
Using `spin_loop_hint` does *not* resolve the LLVM bug.

Further, `spin_loop_hint` is likely to do nothing at all on any sort of `no_std` where there's actually no threading available (as opposed to a `no_std` lib that is then called by `std` using code that manages the threads).

jyn514 commented:
Yeah really what this needs is a `std::hint::make_forward_progress()` intrinsic, but that doesn't exist AFAIK. Maybe you can emulate it with a volatile read/write?

Lokathor commented:
You can.
```rust
let x = 0;
unsafe { (&x as *const i32).read_volatile() };
```
All the "forward progress" you'd ever need.

jyn514 commented:
Nice, that looks about right.

@josephlr either way, this should not have a `spin_loop_hint()` suggestion; I would recommend the volatile read unconditionally since this pops up for both std and no-std.

jyn514 commented:
> In non-no_std code, you usually want to sleep or block, so I think we should not suggest the volatile read there.

Sure, if you want to suggest `thread::sleep(...)` instead that sounds better.

>  For no_std that could work, but personally I'm a bit concerned about suggesting unsafe instead of the (supposedly) safe code of loop {}.

🤷 this is the situation we're in. It's better to have visible unsafe than invisible unsoundness. Right now `loop {}` looks the same to LLVM as `unsafe { unreachable_unchecked() }` which is clearly worse.

ebroto commented:
>It's better to have visible unsafe than invisible unsoundness

I guess you're right. We can always suggest it as a last resort.

Lokathor commented:
no_std has no portable way to sleep. you can't suggest thread::sleep, because there's no such thing.

It's maybe a drag, but as jyn514 said, this is the world we live in.

jonas-schievink commented:
You can also suggest compiling with `-Zinsert-sideeffect` on nightly

therealprof commented:
How about making sure that `spin_loop_hint` always has a side effect? 

The documentation clearly states its intended use:
```
Signals the processor that it is inside a busy-wait spin-loop ("spin lock").

Upon receiving spin-loop signal the processor can optimize its behavior by, for example, saving power or switching hyper-threads.
```

If it cannot be used for an empty loop I'd consider that a bug in `spin_loop_hint`.

BTW: It works just fine for the majority of Rust supported embedded systems (running on  ARM Cortex-M MCUs) by turning into a `yield` instruction.

therealprof commented:
> It's a hint. A hint never changes the correctness (or not) of a program.

The program is correct with or without the hint. However since empty loops are frowned upon and miscompiled by the compiler there should be a proper and official way to declare an intentionally empty loop.

> For example, I do my embedded work on the GBA, and spin_loop_hint does nothing because the ARM7TDMI CPU of a GBA has no built in assembly instruction for a power saving mode like that.

Fair enough, but what would be the issue using a `nop` instead? The `yield` instruction on most Cortex-M MCUs is also just a fancy way to encode a `nop`.

Lokathor commented:
The fact that the codegen will miscompile the program means that the program is *not* correct with or without the hint.

The *intended* semantics of `loop{}` is "that's fine" the *actual* semantics of `loop{}` is "LLVM will burn you". We have to live by the semantics we get.

The official, unstable, way to handle this problem is the `-Zinsert-sideeffect` flag that `jonas-schievink` mentioned.

The *stable* way to handle this is to volatile read a stack variable, like I said.

> Fair enough, but what would be the issue using a nop instead?
Well, because that's not a loop.

therealprof commented:
> The fact that the codegen will miscompile the program means that the program is _not_ correct with or without the hint.
> The _intended_ semantics of `loop{}` is "that's fine" the _actual_ semantics of `loop{}` is "LLVM will burn you". We have to live by the semantics we get.

You're sort of contraticting yourself. A program following the _intended_ semantics is a correct program. Linting about difference actual vs intended semantics and telling the user how to bridge the gap is exactly the job of clippy.

> The official, unstable, way to handle this problem is the `-Zinsert-sideeffect` flag that `jonas-schievink` mentioned.
> The _stable_ way to handle this is to volatile read a stack variable, like I said.

 Empty loops are an essential part of any embedded application (and quite a few other applications, too); telling a user to use a weird unrelated unsafe magic or go unstable is not a solution.

> > Fair enough, but what would be the issue using a nop instead?

> Well, because that's not a loop.

This is about preventing the loop from being turned into mush by creating a side effect that forces the compiler to keep it. Of course the `nop` goes into `loop` and does not replace it...

Lokathor commented:
> Empty loops are an essential part of any embedded application (and quite a few other applications, too); telling a user to use a weird unrelated unsafe magic or go unstable is not a solution.

Anything else is UB.

Any volatile isn't "weird unrelated unsafe magic", every embedded developer needs to understand when to use volatile.

therealprof commented:
> > Empty loops are an essential part of any embedded application (and quite a few other applications, too); telling a user to use a weird unrelated unsafe magic or go unstable is not a solution.
 
> Anything else is UB.

It's not supposed/documented to be. And at the moment there's also no lint for it which is what this issue is all about. People are regularly running into problems with miscompilations due to this.

> And volatile isn't "weird unrelated unsafe magic", every embedded developer needs to understand when to use volatile.

The matter of fact is a lot of people don't know it and they shouldn't have to (if we did our job right and abstracted all of that away which we do!). Requiring people to know things like that is a substantial hazard to Rust adoption and requiring such a magic block of unsafe codeit in **every application** is just ridiculous.

josephlr commented:
Looking at all the comments it seems like we will need to cover multiple use-cases including `no_std` applications and `no_std` libraries. I also agree with the above comments that our advice should make sense _even if we didn't have this LLVM bug_. This seems reasonable, as a hot CPU deadloop is almost always bad style.

Here would be my plan:
  - Remove references to `spin_loop_hint`, as it doesn't (yet) do the right thing for all targets.
    - Once `core` is fixed to make `spin_loop_hint` have a side-effect on all platforms, we recommend that.
  - Update the `no_std` prompt to say something like:
    ```
    empty `loop {}` detected. You may want to either use `panic!()`
    or add a platform-specific yield intrinsic to the loop body.
    ```
    - This handles the `no_std` library case (where libraries should panic instead of deadlooping)
    - This also handles the `no_std` application case (where a binary should do _something_ in it's panic handler other than deadlooping).
  - Update the docs to mention various remediations:
    - Mention that most deadloops outside of the panic handler should panic instead of deadlooping
    - Call out the LLVM bug
    - Calling some platform intrinsic
    - Using `-Zinsert-sideeffect`
    - spin_loop_hint

jyn514 commented:
> a hot CPU deadloop is almost always bad style

This is exactly why I think this should _not_ treat this as if the LLVM bug should not exist. Linting on `loop {}` because it burns up the CPU is performance issue; linting on it because it causes a miscompilation is a correctness issue. If we lump them together I think people will just ignore the lint.

josephlr commented:
> Linting on `loop {}` because it burns up the CPU is performance issue; linting on it because it causes a miscompilation is a correctness issue.

That makes sense. Are you suggesting changing the lint category from "Style" to "Correctness"?

Lokathor commented:
@therealprof
> It's not supposed/documented to be.

But it **is** how things are. There's an open bug about it. No amount of "it shouldn't be that way" will change the facts.

Sometimes Rust just *isn't good yet* and you have to wait for it to improve. If you want this fixed, help patch LLVM to allow side-effect free loops, or help move `insert-sideeffect` to be on by default and stable.

Clippy should give accurate information about the current state of the compiler. If the compiler improves later then the lint can be updated later to match the improvement.

@josephlr 
> Once core is fixed to make spin_loop_hint have a side-effect on all platforms, we recommend that.

Is there even an issue about this ever happening? Because I don't think this will otherwise happen.

josephlr commented:
> Is there even an issue about this ever happening? Because I don't think this will otherwise happen.

I'm working on a patch for `libcore` right now.

therealprof commented:
> But it **is** how things are. There's an open bug about it. No amount of "it shouldn't be that way" will change the facts.

Not sure where you got this notion from. We're having this discussion on the clippy repository because we want to tell people about the facts and how to remedy them (in a sane way).
 
> or help move `insert-sideeffect` to be on by default and stable.

I am not convinced that this is a solution, at best it's a workaround that has the potential to pessmize generated code even further.

> Clippy should give accurate information about the current state of the compiler. If the compiler improves later then the lint can be updated later to match the improvement.

I fully agree.

jyn514 commented:
> Are you suggesting changing the lint category from "Style" to "Correctness"?

flip1995 commented:
After reading all the comments in this thread and most of the comments in https://github.com/rust-lang/rust/issues/28728 I conclude:

1. The lint should be `correctness`, since `loop {}` can produce UB. We should also state this in the lint error message, because:
    > Clippy should give accurate information about the current state of the compiler. If the compiler improves later then the lint can be updated later to match the improvement.

2. It should also lint for `no_std` crates

3. Suggestion is kind of hard, since there is no good solution, that works on all platforms. So we could do the following:
    a. Mention every ~~solution~~ workaround in the lint message (meh)
    b. Try to detect the platform and suggest the best workaround (how?, which?)
    c. Just point to the documentation and write up a summary of the workarounds there

3.a. seems to be too much (useless) information in the lint message.
3.b. seems hard to do. @josephlr suggestion

> Update the `no_std` prompt to say something like:
> ```
> empty `loop {}` detected. You may want to either use `panic!()`
> or add a platform-specific yield intrinsic to the loop body.
> ```
> 
> * This handles the `no_std` library case (where libraries should panic instead of deadlooping)
> * This also handles the `no_std` application case (where a binary should do _something_ in it's panic handler other than deadlooping).

may work for `no_std`, but I wouldn't know what to do and would have to read some documentation anyway.

So I would go with 3.c. What workarounds should we suggest? (please add to the list, if I forgot something)

- Use `-Zinsert-sideeffect`. From reading https://github.com/rust-lang/rust/issues/28728, this seems to me to be the correct solution for this, but requires unstable features, which may not be possible for some people
- Use some volatile read operation. In https://github.com/rust-lang/rust/issues/28728 it was suggested to add `unsafe {llvm_asm!("" :::: "volatile")}`, which prevents this bug, but again needs the unstable feature `llvm_asm`. The `read_volatile()` produces unnecessary code though (including an allocation).
- > Calling some platform intrinsic
    
    I'm also ok with that, with a little more explanation and best case with an example. Is there such a thing for all platforms?
- `spin_loop_hint` and similar functions, with the hint, that they may not work on all platforms.

jyn514 commented:
> spin_loop_hint and similar functions, with the hint, that they may not work on all platforms.

These won't work on *any* platform, they don't involve forward progress. @josephlr suggested changing them so that they emit some sort of intrinsic to LLVM, but there's no guarentee that change is going to be accepted; personally, it seems like solving this in the wrong place to me.

jyn514 commented:
>  Calling some platform intrinsic

I'm also dubious this will work, precisely because the bug is in LLVM and not the platform.

I think suggesting `-Z insert-sideeffect` on nightly and `read_volatile` elsewhere is unfortunately the best clippy can do currently.

josephlr commented:
@jyn514 @flip1995 I opened https://github.com/rust-lang/rust/pull/77924 to make `spin_loop_hint` do what we want on all platforms (which will make it much easier to give good recommendations to users).

> I'm also dubious this will work, precisely because the bug is in LLVM and not the platform.

Here I was mainly thinking of my binary crates, where I call [`x86_64::instructions::hlt`](https://docs.rs/x86_64/0.12.2/x86_64/instructions/fn.hlt.html) in the panic handler. Obviously, that doesn't work for everyone. `cortex_m` users might call [`delay`](https://docs.rs/cortex-m/0.6.3/cortex_m/asm/fn.delay.html), [`wfi`](https://docs.rs/cortex-m/0.6.3/cortex_m/asm/fn.wfi.html), or [`nop`](https://docs.rs/cortex-m/0.6.3/cortex_m/asm/fn.nop.html).

> I think suggesting `-Z insert-sideeffect` on nightly and `read_volatile` elsewhere is unfortunately the best clippy can do currently.

Ya I think the docs will basically say:
  - Don't `loop { }` because it can be UB and wastes CPU
  - If you can, `panic!` instead. This works everywhere and is better than just freezing up. Libraries should almost always do this.
  - Alternatively, on `std` targets:
    - Call `thread::sleep` or `thread::yield_now`
  - Alternatively, on `no_std` targets:
    - Call a target-specific intrinsic to sleep the thread (I'll give `x86_64::hlt` and a `cortex_m::wfi` as examples).
    - Call `sync::spin_loop_hint` (we would only recommend this if my change to `libcore` is accepted)
  - Finally, if you just want to work around the LLVM bug (but not the CPU issue):
    - For nightly, compile your code with `-Z insert-sideeffect`.
    - For stable, do a `read_volatile` on some local variable.

Lokathor commented:
I think "an allocation" is over-stating it. your function's stack will be a single value bigger than normal, pretty trivial. It also inserts a single instruction into the loop instead of 0 instructions in the loop, which is again a fairly trivial difference, since the speed of the empty loop sure doesn't matter.

josephlr commented:
@jyn514 @flip1995 I updated this PR to:
  - Make the lint a "correctness" lint
  - Improve the error messages
  - Improve the documentation as I laid out in https://github.com/rust-lang/rust-clippy/pull/6162#issuecomment-708062490

PTAL. Note that this will now make `empty_loop` a deny-by-default lint.

flip1995 commented on clippy_lints/src/loops.rs:
```diff
@@ -293,18 +293,48 @@ declare_clippy_lint! {
-    /// anything. Think of the environment and either block on something or at least
-    /// make the thread sleep for some microseconds.
+    /// **Why is this bad?** Due to [an LLVM codegen bug](https://github.com/rust-lang/rust/issues/28728)
+    /// these expressions can produce undefined behavior. Also, these busy loops
```
```suggestion
    /// these expressions can produce undefined behavior (in release mode). Also, these busy loops
```

jyn514 replied:
I would not say 'can produce undefined behavior', I'd say 'can be miscompiled'. The user hasn't done anything wrong, the compiler is interpreting their code incorrectly.

flip1995 replied:
I'm good with both or a combination "can be miscompiled, producing undefined behavior"

flip1995 commented on clippy_lints/src/loops.rs:
```diff
@@ -293,18 +293,48 @@ declare_clippy_lint! {
+    ///
+    /// If you just care about fixing the LLVM bug (and don't care about burning
+    /// CPU), you can:
+    ///   - On nightly, compile your code with `-Z insert-sideeffect`
```
Should we also suggest the `llvm_asm!` version here for nightly? That way the user won't have to set any compiler flags and just has to change some code. Also the `llvm_asm!` version is a bit better than the `read_volatile` version, since it doesn't insert any code. (Even though the inserted code doesn't really make a difference performance wise or in any other way, since, as you stated, the program will be trapped in an infinite loop)

flip1995 replied:
Isn't `asm!` deprecated and got replaced with `llvm_asm!` or something?

`-Zinsert-sideeffect` isn't necessarily perma unstable and may become the default behavior sometime in the future, IIUC.

josephlr replied:
> Isn't `asm!` deprecated and got replaced with `llvm_asm!` or something?

The old `asm!` got renamed to `llvm_asm!`. A new `asm!` was added (which is way better) that has a path to stablization.

> `-Zinsert-sideeffect` isn't necessarily perma unstable and may become the default behavior sometime in the future, IIUC.

It becoming the default behavior is actually my concern (as then the flag would be removed).

flip1995 replied:
https://github.com/rust-lang/rfcs/pull/2843 / https://github.com/rust-lang/rust/issues/70173

Ah `asm!` should be reimplemented, so it can become stable sometime in the future. Alright, in that case we should suggest to use `asm!`

therealprof replied:
> Isn't `asm!` deprecated and got replaced with `llvm_asm!` or something?

Indeed. `llvm_asm!` is the new thing.

> `-Zinsert-sideeffect` isn't necessarily perma unstable and may become the default behavior sometime in the future, IIUC.

The `-Z` flag is most certainly perma unstable, it may resurface under a `-C` flag or become enabled by default (which I really do not hope because it's just a bad and costly workaround), so whatever the stable long term solution will be, it's certainly not a `-Z` flag.

flip1995 replied:
> The `-Z` flag is most certainly perma unstable, it may resurface under a `-C` flag

Oh yeah, `-Z` flags are of course perma unstable, oops. Time to sleep 🛌

josephlr replied:
I changed the recommendation to use an empty assembly statement `asm!("")`. It isn't marked `pure`, so this therefore a sideeffect.

flip1995 reviewed (approved):
I'm overall satisfied with the changes, but would still appreciate a second pair of eyes, possibly with more `no_std` experience than me.

ebroto commented on tests/ui/empty_loop_no_std.rs:
```diff
@@ -10,11 +10,15 @@ use core::panic::PanicInfo;
-    loop {}
+    // This loop should not cause a warning.
+    loop {
+        core::sync::atomic::spin_loop_hint();
```
Should this be updated to match one of the current suggestions (`read_volatile`/`llvm_asm!`)?

josephlr replied:
I'll change it to `read_volatile`.

ebroto reviewed (approved):
LGTM modulo flip1995/jyn514 comments

therealprof commented:
I concur with the PR after the requested changes.

jyn514 commented on clippy_lints/src/loops.rs:
```diff
@@ -502,14 +532,16 @@ impl<'tcx> LateLintPass<'tcx> for Loops {
-                     `std::thread::sleep(..);` to the loop body.",
-                );
+            if block.stmts.is_empty() && block.expr.is_none() {
+                let msg = "empty `loop {}` can cause undefined behavior and wastes CPU cycles";
```
```suggestion
                let msg = "empty `loop {}` is currently miscompiled and wastes CPU cycles";
```

josephlr commented:
@flip1995 @ebroto @jyn514 All comments addressed. I think this is ready for merging.

I hope there won't be too much disruption from this lint going to deny-by-default.

flip1995 commented on clippy_lints/src/loops.rs:
```diff
@@ -293,18 +293,53 @@ declare_clippy_lint! {
+    ///
+    /// If you just care about fixing the LLVM bug (and don't care about burning
+    /// CPU), you can:
+    ///   - On nightly, insert a non-`pure` `asm!` statement. For example:
```
I would still include the hint to `-Zinsert-sideeffect`

josephlr replied:
Added

flip1995 reviewed (approved):
Can you please squash the commits?

josephlr commented:
> Can you please squash the commits?

Done, but the build is still broken until #6178 is merged.

flip1995 commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit 012413d2fc392f9a064576553349d0d42c199427 has been approved by `flip1995`

<!-- @bors r=flip1995 012413d2fc392f9a064576553349d0d42c199427 -->
<!-- homu: {"type":"Approved","sha":"012413d2fc392f9a064576553349d0d42c199427","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit 012413d2fc392f9a064576553349d0d42c199427 with merge 33ec6df7874ac08a170e402581572ce0a8015f3d...
<!-- homu: {"type":"BuildStarted","head_sha":"012413d2fc392f9a064576553349d0d42c199427","merge_sha":"33ec6df7874ac08a170e402581572ce0a8015f3d"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

flip1995 commented:
@bors retry

bors commented:
:hourglass: Testing commit 012413d2fc392f9a064576553349d0d42c199427 with merge a280d3b1d321605160f9fa0a41ee415a8a1d602d...
<!-- homu: {"type":"BuildStarted","head_sha":"012413d2fc392f9a064576553349d0d42c199427","merge_sha":"a280d3b1d321605160f9fa0a41ee415a8a1d602d"} -->

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/1265100337)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/runs/1265100337","builder_name":"checks-action_test"} -->

flip1995 commented:
@bors retry (please?)

bors commented:
:hourglass: Testing commit 012413d2fc392f9a064576553349d0d42c199427 with merge 53a59e5a2cf510e51cc800655bce6d2ca03f147a...
<!-- homu: {"type":"BuildStarted","head_sha":"012413d2fc392f9a064576553349d0d42c199427","merge_sha":"53a59e5a2cf510e51cc800655bce6d2ca03f147a"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

jyn514 commented:
So, after all that, https://github.com/rust-lang/rust/pull/77972 just landed. Maybe we no longer need to sound the alarm on this?

ebroto commented:
IMO we could keep the more detailed documentation regarding the case of burning CPU, but I think the lint should go back to warn-by-default and ignore `no_std`

bors commented:
:hourglass: Testing commit 012413d2fc392f9a064576553349d0d42c199427 with merge 6c4257af75b778675cf1fb79bcea09e2b13039a6...
<!-- homu: {"type":"BuildStarted","head_sha":"012413d2fc392f9a064576553349d0d42c199427","merge_sha":"6c4257af75b778675cf1fb79bcea09e2b13039a6"} -->

jyn514 commented:
In that case someone should r- before bors merges 😆

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

ebroto commented:
@bors r-

(Not sure what triggered that last build 🤔)

josephlr commented:
> IMO we could keep the more detailed documentation regarding the case of burning CPU, but I think the lint should go back to warn-by-default and ignore `no_std`

@ebroto @jyn514 I've updated this PR to do just this. But I've added a TODO to address #6161. I think we would ideally have this still trigger for `no_std` crates (as it's still bad style, in general). If we could somehow avoid firing this lint for `#[panic_handler]`, we could enable it everywhere else.

But this sort of discussion can be continued in #6161.

flip1995 commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit 3807634a470b572303d95feb8a5db273c7cea4af has been approved by `flip1995`

<!-- @bors r=flip1995 3807634a470b572303d95feb8a5db273c7cea4af -->
<!-- homu: {"type":"Approved","sha":"3807634a470b572303d95feb8a5db273c7cea4af","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit 3807634a470b572303d95feb8a5db273c7cea4af with merge fd62c180fef3279925934a317cd4dd3ccecff8dc...
<!-- homu: {"type":"BuildStarted","head_sha":"3807634a470b572303d95feb8a5db273c7cea4af","merge_sha":"fd62c180fef3279925934a317cd4dd3ccecff8dc"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/1304538570), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/1304537348), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/1304574701)
Approved by: flip1995
Pushing fd62c180fef3279925934a317cd4dd3ccecff8dc to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/1304574701","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/1304538570","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/1304537348"},"merge_sha":"fd62c180fef3279925934a317cd4dd3ccecff8dc"} -->
